### ISSUE TITLE
Deploy new search date code

### ISSUE THREAD

**Issue by sean-wood** [FreeUKGen/MyopicVicar#969]
In #939 we refactored the search date code. That needs to be deployed to production.

Effort is not great but elapse time may be several days as all search_records wll need updating.

**Comment by Captainkirkdawson** [comment]
Have deployed to production. currently evaluating performance aspects

**Comment by Captainkirkdawson** [comment]
Performance impact is apparent on full throttle update. Impact is with slaves updating.
Continuing with a .1 sec throttle. Still an impact but appears acceptable.

**Comment by Captainkirkdawson** [comment]
At this rate it will take 40+ days to complete. There has to be a better way.

**Comment by Captainkirkdawson** [comment]
Stopped rebuild while we rethink the rebuild process.

**Comment by edickens** [comment]
Should we have a system whereby we can take out one server and
    rebuild that offline, then just copy across the result?
    
    Just an outsider thinking how he would have done it a few years ago!
    
    E
    
      On 08/10/2016 16:43, Kirk Dawson wrote:
    
    
      Stopped rebuild while we rethink the rebuild process.
      —
        You are receiving this because you are subscribed to this
        thread.
        Reply to this email directly, view
          it on GitHub, or mute
          the thread.
      
        
          
          
        
        
      
      {"api_version":"1.0","publisher":{"api_key":"05dde50f1d1a384dd78767c55493e4bb","name":"GitHub"},"entity":{"external_key":"github/FreeUKGen/MyopicVicar","title":"FreeUKGen/MyopicVicar","subtitle":"GitHub repository","main_image_url":"https://cloud.githubusercontent.com/assets/143418/17495839/a5054eac-5d88-11e6-95fc-7290892c7bb5.png","avatar_image_url":"https://cloud.githubusercontent.com/assets/143418/15842166/7c72db34-2c0b-11e6-9aed-b52498112777.png","action":{"name":"Open in GitHub","url":"https://github.com/FreeUKGen/MyopicVicar"}},"updates":{"snippets":[{"icon":"PERSON","message":"@Captainkirkdawson in #969: Stopped rebuild while we rethink the rebuild process."}],"action":{"name":"View Issue","url":"https://github.com/FreeUKGen/MyopicVicar/issues/969#issuecomment-252431520"}}}

**Comment by Captainkirkdawson** [comment]
We had hoped that the use of document locking in wired tiger would allow us to rebuild a collection on the live system. This clearly is not the solution. Likely due to the index updates that are associated with each record change. 
We need to consider how to rebuild a collection without indexes and then do a bulk rebuild of the indexes when the collection is rebuilt. This would need us to do the rebuild off line and then deploy server by server. Updates by members would likely need to be disabled while this process was undertaken.

**Comment by Captainkirkdawson** [comment]
Eric
Your comment is how the new database software wired tiger was deployed. 
For the most part it was invisible to the end user and with the exception of the update to the master invisible to Members. 
It definitely looks like we will have to adopt that approach for this deployment. Every search record is being changed. Hence the on-line update is simply too demanding. 

For this to work we will, I think, have to lock out Members for a significant period

**Comment by edickens** [comment]
I will have to think about promoting WinFreeREG which Mike has ready to 
go.  The reason is that it will not work without access to the server to 
get Place and Church names.  E

**Comment by Captainkirkdawson** [comment]
Options paper
https://docs.google.com/document/d/1uwGXKex4aSFsX9xiRXYcVnGTmF1aHE_k5CwajW2nMtU/edit#

**Comment by Captainkirkdawson** [comment]
Tested Option A steps 4-6 (Drop ans recreate search records on vervet. Taking 60ms/record. Found we were looking up search record for a freereg1_csv_entry without an index.

**Comment by Captainkirkdawson** [comment]
Rewrote the update to have a force create mode. Retested Option A steps 4-6 (Drop ans recreate search records) on vervet. Running single creation task now down to 7ms/record. Estimated time on vervet 3 days for full create

**Comment by Captainkirkdawson** [comment]
Can run 3 create tasks in parallel. One each on ba,ma and bu. Each task marginally slower. Total cpu rate on vervet up to 70%. Estimated completion time now 1 day on vervet.
Perhaps we should consider doing the rebuild on vervet and syncing a new cluster from there Would likely mean locking members out for 2 days. Written up as option C

**Comment by Captainkirkdawson** [comment]
A decision on the approach to be taken is URGENT if we are to use the current FR window.  I will be unavailable for the next 2 windows in November and December.

**Comment by dougkdev** [comment]
They all look like very reasonable solutions.  I would say go with
whichever you are most comfortable with and feel is least likely to
encounter unexpected issues.

**Comment by dougkdev** [comment]
Since you and Ben are more experienced with mongodb than I am, I guess my
only additional question would be whether the eventual consistency setting
affects the assumption that all of the secondary instances are up to date
when switching back and forth between which server is being used as primary
or being updated, and whether there is any risk of getting into an
inconsistent state of data?

**Comment by benwbrum** [comment]
Other than the likelihood that we won't have the code in place this week,
what are your concerns about D? We could be creating un-indexed collections
without interfering with anything currently running.

Ben

**Comment by Captainkirkdawson** [comment]
I do not really fully understand option D nor how it is significantly different from option B. Regardless  it uses something we have not tried ie the use of ( store_in :collection => ‘new collection name’) followed by renaming of collections. Renaming of indexes does not appear to be possible from my reading. The use of an untried technique is why I also dislike option B. 
In option C we would not bother to create indexes as they have to be recreated on the replica servers anyway.
In my last note to the tech list I summarized it as follows.

The main difference between A and C is that:

In C we have 2 copies of the database. The old and the new. Remember the wired tiger databases are much smaller. We are talking of 30-40 GB each compared with the original ones that were 180GB. We switch when all are ready and the recreation is done on vervet thereby imposing no additional load on colobus.

In A we only ever have 1 database on a server. The operational cluster has a reducing number of members until we switch at the minimum of 2 and then an increasing number as we add members. It has the potential for poorer service. It imposes a heavy cpu load on colobus. This process does not require an initial  database clone to vervet nor an initial sync of colobus. Hence with take about 12 hours less time.

D is a variation on B and I have no experience with either

So my inclination is to go with option C

However I am willing to be convinced that it could be the better way to go. If so then you could perhaps take in on during the November window. If you have the time

**Comment by Captainkirkdawson** [comment]
Interesting a collection rename removes the original collection; perhaps that should have been expected

**Comment by Captainkirkdawson** [comment]
Have now completed testing the collection rebuild with vervet. test3 is now running with a fully converted search record collection. Searches are performing extremely fast with and without date bounds, witnesses are located and emendation versions of names are included.
Observations.
The rebuild involved cloning a copy of the database from colobus to vervet 6hrs
Rebuild of the collection on vervet took 30hrs with 3 update tasks running simultaneously.  Reported cpu utilization by New Relic was 250%. Hence expect that 3 cores of the cpu were in use. There was a FreeCEN2 update running at the same time. Vervet has a faster cpu than colobus and had no competition for cpu or io with the operational application
Building the indexes was carried out in the background index by index with the collection being read for each index and resulted in large indexes and took 13 hrs. Building of indexes in the foreground would have locked the database.
A clone of the database takes 6 hrs. A clone of the search records collection takes 3 hrs. A clone results in the indexes being built in the foreground and using a bulk method where the collection is read on once. This creates very much smaller indexes. 

Conclusion.
Rebuilding and cloning of the database or collection to production servers can be done from vervet with no impact on production if the receiving server is taken out of production during the clone  While not confirmed experimentally doing the rebuild and cloning from colobus will have an impact on production due to high cpu usage and database locking and should be undertaken with colobus removed from production and cloning.

**Comment by Captainkirkdawson** [comment]
Ben wrote:
Kirk and I just had a good discussion of technical and process options.  Here's a brief summary of what we think is the best approach:

1) Postpone the rebuild until the last week of October/first week of November, when I'll be around full-time for a week overlapping with Kirk (and a maintenance window we might grab!), and the new developer may be on staff.
2) Prepare messaging to members that they will not be able to upload files during a week-long window.
3) Kirk transfers information about cloning databases, locking out members, and other production operations to me and the new hire, so that he isn't the only one who knows where the car keys are.
4) We do the rebuild on Vervet--following Option C--to minimize impact on the production systems and set a precedent for intensive operations happing somewhere other than production.

We're both pretty happy with this approach, and may flesh out the details later.

**Comment by Captainkirkdawson** [comment]
As Ben has noted we have agreed on a process and technical approach that addresses our respective concerns and should set a good precedent for FreeCEN and any other application based on Mongodb requiring to conduct a major rebuild or an initial deployment. It also addresses my concerns about there being an undue dependence on myself and not using Ben for "grunt" work when his skills are needed elsewhere.

It is clear that we have run out of time to do the rebuild before the FreeCEN window opens up on Monday given that we have NOT communicated to our transcribers and coordinators that the system will be closed to them for several days and the reasons/benefits of what is being proposed. Delaying until after the next FreeBMD update gives us time to communicate with our Members so that they may prepare for the interruption. 

We do not expect that there will be an impact on the researcher community by following this planned approach.

1) We would like to request from Lemon use of the next maintenance slot during the week of October 31- November 4 as both Ben and I can be available then. If we are successful in the deployment we could offer for his use the FreeREG week which immediately follows. There would be no need for any exchange with FreeCEN.
2) I will document in detail the technical approach we plan to use that is based on option C and the experience in rebuilding the search records collection on test3 as well as the deployment of WiredTiger. This documentation will be located in the system documentation for FreeREG.
3) I will work with EricD to write a communication to be sent to Coordinators and Transcribers by Pat.

**Comment by Captainkirkdawson** [comment]
Pat has endorsed the approach

**Comment by Captainkirkdawson** [comment]
Have written up a detailed step by step process in 

https://docs.google.com/document/d/1uwGXKex4aSFsX9xiRXYcVnGTmF1aHE_k5CwajW2nMtU/edit#heading=h.lzge2scdf70u

Have walked through steps 1-7 in part.

Message has been sent to all active personnel and placed in the Actions landing page.

**Comment by Captainkirkdawson** [comment]
saki database cloned to vervet. Rebuild underway on vervet

**Comment by Captainkirkdawson** [comment]
Completed by Ben and I. Had hiccup with top down passing up task . Separate story genereated.

**Comment by Captainkirkdawson** [comment]
Have completed a reconciliation of entries/search records on production and there are now no londer duplicated search records (Had to remove 13,000 caused by cross over.
